pkg/storage/read_as_of_iterator.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import "github.com/cockroachdb/cockroach/pkg/util/hlc"
+
+// ReadAsOfIterator wraps a SimpleMVCCIterator and only surfaces the latest
+// valid key of a given MVCC key that is also below the asOf timestamp, if set.
+// Further, the iterator does not surface delete tombstones, nor any MVCC keys
+// shadowed by delete tombstones below the asOf timestamp, if set. The iterator
+// assumes that it will not encounter any write intents.
+type ReadAsOfIterator struct {
+	iter SimpleMVCCIterator
+
+	// asOf is the latest timestamp of a key surfaced by the iterator.
+	asOf hlc.Timestamp
+}
+
+var _ SimpleMVCCIterator = &ReadAsOfIterator{}
+
+// Close closes the underlying iterator.
+func (f *ReadAsOfIterator) Close() {
+	f.iter.Close()
+}
+
+// SeekGE advances the iterator to the first key in the engine which is >= the
+// provided key that obeys the ReadAsOfIterator key constraints.
+func (f *ReadAsOfIterator) SeekGE(originalKey MVCCKey) {
+	// To ensure SeekGE seeks to a key that isn't shadowed by a tombstone that the
+	// ReadAsOfIterator would have skipped (i.e. a tombstone below asOf), seek to
+	// the key with the latest possible timestamp that the iterator could surface
+	// (i.e. asOf, if set) and iterate to the next valid key at or below the caller's
+	// key that also obeys the iterator's constraints.
+	maxTime := f.asOf
+	if f.asOf.IsEmpty() {
```

Comment:
nit: this check isn't necessary, since an empty timestamp sorts before all other timestamps in MVCC keys -- passing `f.asOf` will do the right thing.

Reply:
huh, this is kinda subtle. When you pass a key with an empty timestamp to the underlying iterator's SeekGE, you're right, it will seek to the latest version of that key. So yes, we can remove this check, but only because we call the internal seek before doing any key comparisons. 

We can't, for example, remove this `isEmpty()` check in `advance` because an empty `hlcTimestamp.walltime` defaults to 0. 
```
} else if !f.asOf.IsEmpty() && f.asOf.Less(f.iter.UnsafeKey().Timestamp) {
```

Reply:
Yeah, `MVCCKey` and `Timestamp` order timestamps differently. The former orders them in descending order but with 0 first (which means that 3 is "less than" or "ordered before" 2, but 0 is before 3), while `Timestamp` uses regular ascending order (0,1,2,3...). I suppose you get used to it, but definitely subtle and a bit of a footgun.